### ISSUE TITLE
Mob pathing refinement proposal

### ISSUE THREAD

**Issue by OtianNgocnion** [DarkstarProject/darkstar#2360]
In issue https://github.com/DarkstarProject/darkstar/issues/1835 mobpathing was adjusted, and mobs no longer do the strange dance that they used to do. However, I have a proposal for some refinement. 

Currently, when the server wants to path a mob to its target, it does the following: choose a random point on the circle of radius r centered at the target point x and path to that point. This is done in the function PathAround, line 128 of pathfind.cpp. The random angle is drawn on line 133, and the lastPoint is on lines 135 and 136. 

It's easy to test that this is working as written by simply standing still, pulling a mob, and noticing that it will not path to you, but rather to a random point on that invisible circle of radius distanceFromPoint. 

This behavior is stable, in that it works pretty consistently, but the mobs settle down in a strange way, and still are a little jumpy. So, I propose a refinement:

Scrap the random point thing. Instead, generate a path whose endpoint is the target, and then, starting from the end of the path array, remove points that are within distFromPoint to the target. In english: find a path that goes straight to the target, and then stop before you go within distFromPoint.

**Comment by teschnei** [comment]
Problem: every mob will now stack exactly on one another

**Comment by OtianNgocnion** [comment]
Good point. If they're coming from the same direction, they would stack. Hmmm.  So we need an implementation of: find a path that goes straight to the target, stop before you go within distfromPoint, and don't stand on your buddies. 

I'm just getting started digging into the code... Do the mobs have any collision or stacking avoidance already implemented? 

If not, I would suggest to identify the "ideal" point above, and then add jitter/noise to it within a smaller sector of the circle. In a sense, this would be the average of my suggestion and the current implementation: find a random point on the distFromPoint circle, but within, say, +/- pi/8 from the "ideal" point that I described in the initial post.

**Comment by bendangelo** [comment]
Exactly what kj said. The way it works on retail (from what I have seen) is the mob will run directly towards you, and stop once it's within attacking range. Mobs clumped up will automatically spread around you ever so slightly. I haven't taken the time to implement that because I think the system implemented is acceptable for now.

**Comment by m241dan** [comment]
I will add that the mobs will stack ontop of each other if there is enough of them. I don't know if you guys ever did a cleave party jn abysssa but we'd often position ourselves in corners or against walls and drag the mobs backward along a wall until they stacked ontop of each other for an easy AoE weaponskill.

**Comment by teschnei** [comment]
Yes, but that only happens because they can't move around you (the wall is
in the way)

**Comment by m241dan** [comment]
However, I think that because when you pull a mob, they aren't already stacked. Their most direct path(on average) will not cause them to stack except only slightly with each other. Now, I get this might not be entirely retail but lord let me tell you, the way it is right now is SUPER annoying to play with. I'd rather see the mobs pseudo stack on pull than deal with Niddhogg's and other Wyrm's squirrelly asses when trying to pull them. Or adjusting Tiamat or Jorm after hate resets. No, it's not high priority but it is highly annoying.

**Comment by OtianNgocnion** [comment]
Agreed @ everything that has been said. This is why I was wondering about mob collision detection or stacking avoidance: in retail they *would* all stack up, and then slowly drift apart like little point charges. 

What do you guys think of the directPoint +/- uniformRV[0,pi/8] thing from post 3? This would mellow the current behavior, without the stacking.

**Comment by OtianNgocnion** [comment]
[Just an addendum to say that it's being worked on, with teschnei's issue in mind. Thanks for your feedback!]

**Comment by teschnei** [comment]
My suggestion is that if time is being spent on it, it'd be to just make it
behave as retail. I don't think they already have collision detection, so
there'd have to be some way of checking that, hopefully with as little
iteration over mob lists as possible

**Comment by Deadwing888** [comment]
I'm really glad someone is looking into this, it's actually more annoying than you would think in day-to-day play.

Just like m241dan said it's a more serious problem for wyrms. It took me two full minutes to position Tiamat correctly last night (having to walk the whole time so he maybe doesn't draw us under the map again)

**Comment by m241dan** [comment]
Haha, right? Props to you man, I just shut his draw-in off because sometimes if someone pulled hate it would draw everyone in under ground and then you're basically fubared.

**Comment by bendangelo** [comment]
Ok I'll make it more retail

**Comment by OtianNgocnion** [comment]
Already on it, bendangelo :+1: and happy to make some contributions. Thanks for the comments though.
